keps/sig-apps/0024-ttl-after-finish.md

```diff
@@ -0,0 +1,290 @@
+	// TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
+}
+```
+
+This allows Jobs to be cleaned up after they finish and provides time for
+asynchronous clients to observe Jobs' final states before they are deleted.
+
+
+Similarly, we will add the following API fields to `PodSpec` (`Pod`'s `.spec`).
+
+```go
+type PodSpec struct {
+ 	// ttlSecondsAfterFinished limits the lifetime of a Pod that has finished
+	// execution (either Succeeded or Failed). If this field is set, once the Pod
+	// finishes, it will be deleted after ttlSecondsAfterFinished expires. When
+	// the Pod is being deleted, its lifecycle guarantees (e.g. finalizers) will
+	// be honored. If this field is unset, ttlSecondsAfterFinished will not
+	// expire. If this field is set to zero, ttlSecondsAfterFinished expires
+	// immediately after the Pod finishes.
+	// This field is alpha-level and is only honored by servers that enable the
+	// TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
+}
+```
+
+##### Validation
+
+Because Job controller depends on Pods to exist to work correctly. In Job
+validation, `ttlSecondsAfterFinished` of its pod template shouldn't be set, to
+prevent users from breaking their Jobs. Users should set TTL seconds on a Job,
+instead of Pods owned by a Job.
+
+### User Stories
+
+The users keep creating Jobs in a small Kubernetes cluster with 4 nodes.
+The Jobs accumulates over time, and 1 year later, the cluster ended up with more
+than 100k old Jobs. This caused etcd hiccups, long high latency etcd requests,
+and eventually made the cluster unavailable.
+
+The problem could have been avoided easily with TTL controller for Jobs.
+
+The steps are as easy as:
+
+1. When creating Jobs, the user sets Jobs' `.spec.ttlSecondsAfterFinished` to
+   3600 (i.e. 1 hour).
+1. The user deploys Jobs as usual.
+1. After a Job finishes, the result is observed asynchronously within an hour
+   and stored elsewhere.
+1. The TTL collector cleans up Jobs 1 hour after they complete.
+
+### Implementation Details/Notes/Constraints
+
+#### TTL Controller
+We will add a TTL controller for finished Jobs and finished Pods. We considered
+adding it in Job controller, but decided not to, for the following reasons:
+
+1. Job controller should focus on managing Pods based on the Job's spec and pod
+   template, but not cleaning up Jobs.
+1. We also need the TTL controller to clean up finished Pods, and we consider
+   generalizing TTL controller later for custom resources. 
+
+The TTL controller utilizes informer framework, watches all Jobs and Pods, and
+read Jobs and Pods from a local cache.
+
+#### Finished Jobs
+
+When a Job is created or updated:
+
+1. Check its `.status.conditions` to see if it has finished (`Complete` or
+   `Failed`). If it hasn't finished, do nothing. 
+1. Otherwise, if the Job has finished, check if Job's 
+   `.spec.ttlSecondsAfterFinished` field is set. Do nothing if the TTL field is
```

Comment:
Since both steps merely check the fields of the Job, the order is interchangeable. We don't need to do any further computation if either the job hasn't finished or its ttl it's set.

Reply:
I think there is an important distinction here. `.spec.ttlSecondsAfterFinished` is the field that allows this controller to process those jobs. This is what discriminates if a Job is added to a queue or not and we should be very explicit about this. No other action should be taken on a Job if this field is not set. Logically it doesn't make sense to check a status for a Job is you shouldn't be managing it at all. Controllers/code can evolve an we should structure it well from the start even though now the read-only actions don't cause issues.

Reply:
To clarify, Jobs are only added to the workqueue (for cleanup) when they (1) finished (2) have a TTL set. No action will be taken if not both conditions are met.